### PR TITLE
fix(Core/Services): Re-check all achievement criterias on faction cha…

### LINKED ISSUE #12559: Faction Change loses progress on Loremaster from Neutral quest
### Current Behaviour

When completing neutral quest (e.g. Wastewander Justice from Gadgetzan), then switching faction, it will not count this quest towards your total goal

If you did too many  of them, your character is locked out.

### Expected Blizzlike Behaviour

It should count towards total goal.

### Source

...!

### Steps to reproduce the problem

1. .q add 1690
2. .q complete 1690
3. .go quest ender 1690
4. check achievements for kalimdor loremaster see 1/xxx
5. faction change
6. check achievements for kalimdor, see 0/xxx

### Extra Notes

image

image

Alliance ONLY quest will count towards the Total quest done achievement after faction changing to horde, too which doesnt seem right.

### AC rev. hash/commit

07c043552aa833bd101ffc0eb2ed71594fdea773

### Operating system

W11x64

### Custom changes or Modules

none

### PR DESCRIPTION
**Pull request by UltraNix**
…nge.

Fixes #12559

## Issues Addressed:

- Closes #12559

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

`.q add 1690`
`.q complete 1690`
`.go quest ender 1690`
check achievements for kalimdor loremaster see 1/xxx
faction change
check achievements for kalimdor, see 1/xxx

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 1]
```diff
@@ -2541,6 +2541,12 @@ void WorldSession::HandleCharFactionOrRaceChangeCallback(std::shared_ptr<Charact
         }
     }
 
+    // Re-check all achievement criterias
```
Comment (pklloveyou) [inline src/server/game/Handlers/CharacterHandler.cpp:2544]: if (recvData.GetOpcode() == CMSG_CHAR_FACTION_CHANGE)
better？

Comment (UltraNix) [reply src/server/game/Handlers/CharacterHandler.cpp:2544]: No. On race change should be the same.

**Comment by pklloveyou** [comment]
I just tried that this code is not required for race change, but only for camp change. Race change will not cause errors

**Comment by UltraNix** [comment]
@pklloveyou It is required for other stuff (e.g. race-specific quests rewarded or other race-specific sutff).

**Comment by pklloveyou** [comment]
OK, thank you for your answer

**Comment by pklloveyou** [comment]
Tested

**Review by IntelligentQuantum — APPROVED** [review]

**Pull request merged**